src/rgw/rgw_redis_scripts.lua

```diff
@@ -0,0 +1,231 @@
+    if not redis.call('LPUSH', name, value) then
+        return formatResponse(-lerrorCodes.ENOMEM, "Not enough memory", "")
+    end
+    return formatResponse(0, "", "")
+end
+
+--- Remove an item from the reserve queue
+--- @param keys table A single element list - queue name
+--- @return number 0 if the item is removed from the queue, -lerrorCodes.ENOENT if the queue is empty
+local function unreserve(keys)
+    local name = "reserve:" .. keys[1]
+
+    local value = redis.call('RPOP', name)
+    if not value then
+        return formatResponse(-lerrorCodes.ENOENT, "Queue is empty", "")
+    end
+    return formatResponse(0, "", "")
+end
+
+--- Commit message to the queue
+--- @param keys table A single element list - queue name
+--- @param args table A single-element list - message
+--- @return number 0 if the message is committed to the queue
+local function commit(keys, args)
+    local name = "queue:" .. keys[1]
+    local message = args[1]
+
+    unreserve(keys)
+    if not redis.call('LPUSH', name, message) then
+        return formatResponse(-lerrorCodes.ENOMEM, "Not enough memory", "")
+    end
+    return formatResponse(0, "", "")
+end
+
+--- Abort the message reservation
+local function abort(keys, args)
+    return unreserve(keys)
+end
+
+--- Read a message from the queue
+--- @param keys table A single element list - queue name
+--- @return string message if the message is read from the queue, nil if the queue is empty
+--- @return This does not remove the message from the queue
+local function read(keys)
+    local name = "queue:" .. keys[1]
+    local value = redis.call('LRANGE', name, -1, -1)[1]
+    return formatResponse(0, "", value)
+end
+
+--- Option one
+--- Have a separate read if lock is held
+--- @param keys table A single element list - queue name
+--- @param args table A single element list - cookie
+--- @return pair of number (error code) and string (message)
+local function locked_read(keys, args)
+    local name = "queue:" .. keys[1]
+    local cookie = args[1]
+
+    local assert_lock_keys = {"lock:" .. keys[1]}
+    local assert_lock_args = {cookie}
+
+    local lock_status = assert_lock(assert_lock_keys, assert_lock_args)
+    if lock_status.map.errorCode == 0 then
+        local value = redis.call('LRANGE', name, -1, -1)[1]
+        return formatResponse(0, "", value)
+    end
+    return lock_status
+end
+
+--- Acknowledge the Read
+--- @param keys table A single element list - queue name
+--- @param args table A single element list - cookie
+--- @return number 0 if the message is acknowledged
+local function ack_read(keys, args)
+    local name = "queue:" .. keys[1]
+    local cookie = args[1]
+
+    local assert_lock_keys = {"lock:" .. keys[1]}
+    local assert_lock_args = {cookie}
+
+    local lock_status = assert_lock(assert_lock_keys, assert_lock_args)
+    if lock_status.map.errorCode == 0 then
+        redis.call('RPOP', name)
```

Comment:
is it possible to pop multiple messages from the queue.
can we pass the number of messages to pop as a parameter?

Reply:
Yup, if we want to read and subsequently ack in batches we can do a `LMPOP` with the number of elements to pop.

Reply:
this woudl b great and improve performence as well. especially with batched backends like amqp and kafka, where you can ack multiple notifications at once